# Exploring Hacker News Posts

In this project, we are going to explore a dataset related to Hacker News Posts. The original dataset can be found in [this Kaggle page](https://www.kaggle.com/hacker-news/hacker-news-posts).

However, it has ~300k entries and, for this project, a modification has been done. The new dataset has ~20k entries (after removing posts without comments and sampling the remaining ones) and can be downloaded from the Dataquest page by clicking on "Download" button in the embedded Jupyter notebook. 

## Goals of the project

In particular, we are going to have a look at the posts that begin with:
- **Ask HN**: They are used to ask question to HN community.
- **Show HN**: THey are used to simply show something to HN community.

From these 2 groups we are going to:
1. Analyse which one of these receive more comments on average.
2. Analyse if the time creation affects to the number of comments.

## Description of the dataset

The dataset includes around 20000 posts information, splitted in different rows.

There are 7 columns in each row, with the following information:
- **id**: The unique identifier from Hacker News for the post.
- **title**: The title of the post.
- **url**: The URL that the posts links to, if the post has a URL.
- **num_points**: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes.
- **num_comments**: The number of comments that were made on the post.
- **author**: The username of the person who submitted the post.
- **created_at**: The date and time at which the post was submitted.

In [1]:
from csv import reader

open_file = open("my_datasets/hacker_news_posts/hacker_news.csv", encoding='utf8')
read_file = reader(open_file)
hn = list(read_file)

#Printing the first 5 lines of hn variable, including header
for row in range(5):
    print(hn[row])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']
['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']
['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']
['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


In [2]:
#Splitting the read file in header and content without header
header = hn[0]
hn = hn[1:]

#Checking the result
print("Header:")
print(header)
print("\n")
print("The hn list has ",len(hn)," entries.")
print("First 5 hn rows:")
for row in range(5):
    print(hn[row])

Header:
['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


The hn list has  20100  entries.
First 5 hn rows:
['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']
['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']
['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']
['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']
['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8',

In [3]:
#Dividing content in ask posts, show posts and others
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

#Checking length of the final lists
print("ask_posts list has ",len(ask_posts)," entries.")
print("show_posts list has ",len(show_posts)," entries.")
print("other_posts list has ",len(other_posts)," entries.")

ask_posts list has  1744  entries.
show_posts list has  1162  entries.
other_posts list has  17194  entries.


In [4]:
#Comparing the total number of comments between ask_posts and show_posts

total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print("The average number of comments in ask posts is: ",round(avg_ask_comments,2),".")

total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    
avg_show_comments = total_show_comments / len(show_posts)
print("The average number of comments in show posts is: ",round(avg_show_comments,2),".")

The average number of comments in ask posts is:  14.04 .
The average number of comments in show posts is:  10.32 .


## First results

After comparing the average number of comments in Ask Posts and Show Posts we can see that Ask Posts wins.

In particular, **Ask Posts has 14.04** comments on average while **Show Posts receive 10.32** comments.

From now on, we will work with the Ask Posts list to analyse the relationship between comments on posts and the time they were created.

## Relationship between post creation time and comments

Ask Posts list will be analysed:

In [5]:
import datetime as dt

result_list = []

for row in ask_posts:
    # We will append in the result_list a list with 2 columns:
    # creation date and number of comments (as integer).
    new_element = [row[6],int(row[4])]
    result_list.append(new_element)
    
counts_by_hour = {}
comments_by_hour = {}

#Convert time in datatime object
for element in result_list:
    element[0] = dt.datetime.strptime(element[0],"%m/%d/%Y %H:%M")   

In [6]:
#Parse hour and fill dictionaries
for element in result_list:
    hour = element[0].strftime("%H")
    comment = element[1]
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment

In [7]:
#Calculating the average comments per hour
avg_by_hour = []
for hour in counts_by_hour:
    num_posts = counts_by_hour[hour]
    comments = comments_by_hour[hour]
    average = comments/num_posts
    avg_by_hour.append([hour,average])
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [8]:
#Creating the swaped list of list, to later sort by average number
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour)

#Sorting by average number
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [10]:
#Printing the Top 5 Hours for Ask Posts Comments
print("Top 5 Hours for Ask Posts Comments:")
for i in range(5):
    rank_time = sorted_swap[i][1]
    rank_comments = sorted_swap[i][0]
    dt_time = dt.datetime.strptime(rank_time,"%H")
    print(dt_time.strftime("%H:%M:"),'{:.2f}'.format(rank_comments),"average comments per post")

Top 5 Hours for Ask Posts Comments:
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


## Conclusion

According to the results, the best hour to get more comments in an Ask Post is 15:00h Eastern Time US.

Comparing to our time zone (Central Europe Time), the best time for us to write an Ask Post would be 21:00h.

After checking the Top 5, we can see an increase of more than 50% between the first and the second option.